### PR TITLE
Fixed potential exception due to stale profile UUID

### LINKED ISSUE #54656: LOGICAL_ERROR from SettingsProfilesInfo::getProfileNames()
An undisclosed customer recently ran into this `LOGICAL_ERROR`:

```
Code: 49. DB::Exception: Unable to get profile name for 834b80a5-c274-88dd-9e7b-8e83274cbf85. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000ead79f7 in /usr/bin/clickhouse
1. ? @ 0x000000000914938d in /usr/bin/clickhouse
2. DB::SettingsProfilesInfo::getProfileNames() const @ 0x0000000013ab69cc in /usr/bin/clickhouse
3. DB::SessionLog::addLoginSuccess(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::optional<String>, DB::Context const&, std::shared_ptr<DB::User const> const&) @ 0x00000000144bc5fb in /usr/bin/clickhouse
4. DB::Session::makeQueryContextImpl(DB::ClientInfo const*, DB::ClientInfo*) const @ 0x00000000144aff18 in /usr/bin/clickhouse
5. DB::TCPHandler::receiveQuery() @ 0x000000001567a57c in /usr/bin/clickhouse
6. DB::TCPHandler::receivePacket() @ 0x000000001567040e in /usr/bin/clickhouse
7. DB::TCPHandler::runImpl() @ 0x0000000015667890 in /usr/bin/clickhouse
8. DB::TCPHandler::run() @ 0x000000001567ef59 in /usr/bin/clickhouse
9. ? @ 0x000000000eba8dbc in /usr/bin/clickhouse
10. Poco::Net::TCPServerConnection::start() @ 0x0000000018687df4 in /usr/bin/clickhouse
11. Poco::Net::TCPServerDispatcher::run() @ 0x0000000018689011 in /usr/bin/clickhouse
12. Poco::PooledThread::run() @ 0x0000000018812207 in /usr/bin/clickhouse
13. Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001880fc3c in /usr/bin/clickhouse
14. ? @ 0x00007f05dc96fb43 in ?
15. ? @ 0x00007f05dca01a00 in ?
```

This is basically old issue #35952 for which supposed fix #42641 was made - the problem was super sporadic even back then, so the fix could not be reproduced/verified.

What happens is that during the creation of a session, `SettingsProfilesInfo::getProfileNames()` is called which eventually finds that its members `profiles` and `names_of_profiles` are out of sync. There was some speculation in #42641 how this could happen. Back then, everyone concluded that the only place which modifies (actually: creates) `SettingsProfilesInfo` is `SettingsProfilesCache::substituteProfiles()`. The locking is good, we can exclude races as cause (see the fancy figure in the other bug), I also quickly confirmed that the locking is still okay. `substituteProfiles()` will also not corrupt `SettingsProfilesInfo` in case of OOM or an exception, we are also safe from this side.

But: As I just see, both callers of `substituteProfiles()`

1. `getSettingsProfileInfo()`:
```cpp
info->profiles.push_back(profile_id);
info->profiles_with_implicit.push_back(profile_id);
substituteProfiles(elements, info->profiles_with_implicit, info->names_of_profiles);
```

2. `mergeSettingsAndConstraintsFor()`
```cpp
info->profiles = merged_settings.toProfileIDs();
substituteProfiles(merged_settings, info->profiles_with_implicit, info->names_of_profiles);
```

store something in `profiles` but they don't store something corresponding in `names_of_profiles`. There is a chance that this step corrupts `SettingsProfilesInfo` when `elements` resp. `merged_settings` is empty (as `substituteProfiles()` becomes a no-op then). Also, tbh., the code in `substituteProfiles()` itself also looks a bit strange (iterate the map while modifying it).

@vitlibar - maybe you like to double check (as author of the code)?

### PR DESCRIPTION
**Pull request by Enmk**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed (a rare) exception in case when user's assigned profiles are updated right after user logging in, which could cause a missing entry in `session_log` or problems with logging in.

`SettingsProfilesInfo::profiles` is not updated in bg if any of the profiles assigned to the user change, but `SettingsProfilesInfo::profiles_with_implicit` is.

Update of #42641
kudos @tavplubix  https://github.com/ClickHouse/ClickHouse/pull/42641/files/3d0c07ac5b8f18917f2314474030910176ec7940#r1406196201

Closes: #54656

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 174607c6bf20c800298595de82d24ccb0b894197 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57263/174607c6bf20c800298595de82d24ccb0b894197/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7006219608">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57263/174607c6bf20c800298595de82d24ccb0b894197/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Access/SettingsProfilesInfo.cpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ Strings SettingsProfilesInfo::getProfileNames() const
 {
     Strings result;
     result.reserve(profiles.size());
-    for (const auto & profile_id : profiles)
+    for (const auto & profile_id : profiles_with_implicit)
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Access/SettingsProfilesInfo.cpp:69]: For example, we can fix that inconsistency by adding a new argument `profile_ids` to `substituteProfiles()`:

```
void SettingsProfilesCache::substituteProfiles(
    SettingsProfileElements & elements,
    std::vector<UUID> & profile_ids,
    std::vector<UUID> & substituted_profiles,
    std::unordered_map<UUID, String> & names_of_substituted_profiles) const
{
    profile_ids = elements.toProfileIDs();
    substituted_profiles = profile_ids;
 
   ... (the current body of substituteProfiles() )

    boost::range::remove_erase_if(profile_ids, [](const UUID & profile_id)
    {
        return !substituted_profiles_set.contains(profile_id);
    });
}
```

**Review by vitlibar — APPROVED** [review]

File: src/Access/SettingsProfilesCache.cpp [hunk 2]
```diff
@@ -184,6 +187,11 @@ void SettingsProfilesCache::substituteProfiles(
         names_of_substituted_profiles.emplace(profile_id, profile->getName());
     }
     std::reverse(substituted_profiles.begin(), substituted_profiles.end());
+
+    boost::range::remove_erase_if(profiles, [&substituted_profiles_set](const UUID & profile_id)
```
Comment (rschu1ze) [inline src/Access/SettingsProfilesCache.cpp:191]: Can that be done with some STL built-in instead of boost?

Comment (vitlibar) [reply src/Access/SettingsProfilesCache.cpp:191]:
```
boost::range::remove_erase_if(profiles, func);
```
seems shorter and clearer than
```
profiles.erase(std::remove_if(profiles.begin(), profiles.end(), func), profiles.end());
```

Comment (Enmk) [reply src/Access/SettingsProfilesCache.cpp]: Turns out, there is a relatively new (C++20) [`std::erase_if`](https://en.cppreference.com/w/cpp/container/vector/erase2), which is equivalent to the `boost::range::remove_erase_if`

**Comment by vitlibar** [comment]
> `currentProfiles()` is not working now, see the [test](https://s3.amazonaws.com/clickhouse-test-reports/57263/b634e043d5e64073961bd9cd337f4dbef6615657/integration_tests__asan__%5B3_4%5D/integration_run_parallel4_0.log)

Looks like this code
```
SettingsProfileElements elements;
elements.push_back(SettingsProfileElement{.parent_profile = profile_id});
```
instead of [that line](https://github.com/ClickHouse/ClickHouse/blob/4d27f87be89fa2c3eeb914398334eb9dac0cd32f/src/Access/SettingsProfilesCache.cpp#L228) must help.

**Review by vitlibar — APPROVED** [review]

**Comment by vitlibar** [comment]
Test failures:
- [00191_aggregating_merge_tree_and_final](https://s3.amazonaws.com/clickhouse-test-reports/57263/2a385bc573879a45b2ca8332d4abb91c9b862609/stateless_tests__release__s3_storage__[1_2].html) - see [issue 58337](https://github.com/ClickHouse/ClickHouse/issues/58337)
- [Logical error: not_processed should be empty](https://s3.amazonaws.com/clickhouse-test-reports/57263/2a385bc573879a45b2ca8332d4abb91c9b862609/upgrade_check__asan_.html) - see [issue 58319](https://github.com/ClickHouse/ClickHouse/issues/58319)

**Pull request merged**